Fx(117): examples for 'auto none' in 'contain-intrinsic-size' CSS prop (#28618)
mdn/content: files/en-us/web/css/contain-intrinsic-size/index.md

>   - : The element has no intrinsic size in the given dimension(s).

Reviewer: Above it says:

> Either one or two of the following values may be specified for an element.
> If two values are specified, the first value applies to the width, and the second to the height.
> If a single value is specified, it applies to both width and height.

This is correct only if you assume the two-value specifications, such as "auto length", count as one value. It is jarring to see `contain-intrinsic-size: auto 300px auto 4rem;` above since that has 4 values.

Author: Yes, this no longer applies

Author: Added a reword in d0002a157e8f8c000cb00f0a7857cbdc836d10ad

Reviewer: Good fix

This allows offscreen elements with [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility) to benefit from size containment without developers having to be as precise in their estimates of element size.
The remembered value is not used if the child elements are being rendered (if size containment is enabled, the `<length>` will be used).

In grid and multi column layouts, an explicit height is treated as a stronger command than an implicit content-based height.

> Elements might lay out substantially differently than it would have were it simply filled with content up to that height.

Reviewer: The first sentence doesn't mean anything to me, but the second seems a little more useful. Is it just height that affects layout or would a width setting do something?

I mean **how** might a layout be different with none vs 0 in those cases? My "guess" is that if you specify `none` a cell might fully collapse to nothing - so you lose the affect of  padding (but keep margins). Whereas if you have 0 you get padding and margins still. But that's a guess. I think this would be better with a more concrete example.

Author: Added something in the comment below, linking here also for convenience that I think covers this: https://codepen.io/bsmth/pen/rNoLKzL?editors=1100

This allows offscreen elements with [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility) to benefit from size containment without developers having to be as precise in their estimates of element size.
The remembered value is not used if the child elements are being rendered (if size containment is enabled, the `<length>` will be used).

In grid and multi column layouts, an explicit size is treated differently than implicit content-based height.
Elements might lay out substantially differently than it would have were it simply filled with content up to that height.

> The `auto none` value allows the element to fallback to `contain-intrinsic-size: none` if no remembered value exists, which will allow the element to be laid out as though it had no contents instead of 0px height.

Reviewer: I still don't know _how_ is it different to lay out elements using 0 vs `none`.
The reason I think it matters is that I am still not sure when I would use one or the other.

Am I being stupid? Other than that I think this is excellent.

Author: > I still don't know _how_ is it different to lay out elements using 0 vs `none`. The reason I think it matters is that I am still not sure when I would use one or the other.
> 
> Am I being stupid? Other than that I think this is excellent.

Not at all, this is good to have clear here; I think it's not common or recommended to use 0px sizes for areas/cells in grid and multicol layouts because you will get horrible content overflow (see [codepen](https://codepen.io/bsmth/pen/rNoLKzL?editors=1100)) and depending on what the contents are in the cells (either many or complex nested elements) it will likely completely break page layout.

Reviewer: Thanks, that answers my question. How about this?

```suggestion
The `auto none` value allows the element to fallback to `contain-intrinsic-size: none` if no remembered value exists, which will allow the element to be laid out as though it had no contents. This is almost always preferred to setting 0px as the intrinsic size in grid and multi column layouts, which allows contained text to overflow and may result in unexpected page layout.
```

Author: Looks good, I made one tweak, but this is much better

## Examples

### Using auto value pairs for intrinsic size

> This example demonstrates `contain-intrinsic-size: auto <length>` and `contain-intrinsic-size: auto none` approximating an 'infinite scroll' layout where there are many of the same type of elements repeated vertically.

Reviewer: I like to say what the example actually does. Wasn't sure about utility of mentioning infinite scroll. Maybe:

```suggestion
This example demonstrates `contain-intrinsic-size: auto <length>` and `contain-intrinsic-size: auto none`, using a layout where there are many elements displayed vertically that have both accurate and incorrect intrinsic size estimations.
```

Using `content-visibility: auto` skips rendering elements when they are offscreen, so this property is a good candidate to combine with `contain-intrinsic-size` to improve rendering performance and minimize [reflows](/en-US/docs/Glossary/Reflow).

The `contain-intrinsic-size: auto 500px` value pair tells the browser to use 500px as a kind of 'placeholder' size (width and height) for the element when it is offscreen and the page is being laid out.

> When the element needs to be rendered onscreen (when the user scrolls to it), the browser will perform a render of the element and its contents.

Reviewer: ```suggestion
When the user scrolls to the element and it needs to be displayed, the browser will calculate the actual size of the element and its contents.
```

Once the browser has actual size information for the element, it will remember this size when the element scrolls offscreen again, and use the remembered size for layout calculations instead of the placeholder value.
The benefit is that the browser does not need to repeatedly render the element contents to calculate its size and is especially useful when the contents are complex or depend on network resources or JavaScript.

#### HTML

```html hidden
<p id="auto-none-note">
  Your browser does not support "<code>contain-intrinsic-size: auto none</code
  >".
</p>
<p id="auto-length-note">
  Your browser does not support "<code
    >contain-intrinsic-size: auto &lt;length&gt;</code
  >".

> </p>

Reviewer: You might be able to force a nicer prettier format by playing around with the initial tags. Just a thought.

Author: Added something in d0002a157e8f8c000cb00f0a7857cbdc836d10ad

  height: max-content;
  font-size: 1rem;
  background-color: tomato;
}

@supports (contain-intrinsic-size: auto none) {
  .auto-length.none {
    display: block;
  }
  #auto-none-note {
    display: none;
  }
}
@supports (contain-intrinsic-size: auto 500px) {
  .auto-length {
    display: block;
  }
  #auto-length-note {
    display: none;
  }
}
```

```css
p {
  height: 500px;
  width: 500px;
  border: 4px dotted;
  background: lightblue;
}

.auto-length {
  content-visibility: auto;
  contain-intrinsic-size: auto 500px;
  background-color: linen;
  outline: 4px dotted blue;
}

.large-intrinsic-size {
  /* Setting an inaccurate intrinsic size for the element */
  contain-intrinsic-size: auto 5000px;
  background-color: lightgray;
  outline: 4px dotted red;
}

.small {
  /* This element is a lot smaller than expected */
  height: 100px;
  width: 100px;
}

.none {
  background-color: papayawhip;
  contain-intrinsic-size: auto none;
  outline: 4px dotted red;
}
```

#### Result

- The first two boxes have an intrinsic size that matches their actual size, so as they flow into view, the layout is recalculated but we see no change in the scrollbar or the scroll position.

> - The third and fourth boxes have a huge intrinsic size, so the initial layout that the browser calculated is far too big.
>   The paragraphs in these boxes are much smaller so that it's obvious you've reached a point that forces a drastic layout change.

Reviewer: ```suggestion
- The third and fourth boxes have a huge intrinsic size, so the initial layout that the browser calculated is far too big, and we've made these boxes smaller so that it's obvious when you've reached a point that forces a drastic layout change.
```

  When the third and fourth boxes scroll into view, the size is recalculated, making the box and its parent less tall.
  The effect is that the scroller jumps down the page (we've effectively scrolled further through the box than we'd estimated) and the scroller is longer, because the entire page is less tall than we'd estimated.

- The last boxes have `auto none`, so they have zero estimated size.

>   When they scroll into view the size of the parent increases, the scroller decreases in size and moves up the bar.

Reviewer: ```suggestion
  When they scroll into view the size of the element and its parent are recalculated to be much larger, so the scroller decreases in size and moves up the bar.
```

> 

Reviewer: ```suggestion

After scrolling all the way to the bottom you can subsequently scroll up and down smoothly, because using `content-visibility: auto` saves the actual rendered size of the element for next time it is displayed.

```

## Examples

> ### Using auto value pairs for intrinsic size

Reviewer: Does the formal syntax allow for this new case? I'm not seeing it.

![image](https://github.com/mdn/content/assets/5368500/08a0ed03-b07b-46d6-acdf-3a6a1fdf95ff)

Reviewer: I'm not sure. REview is just a "this struck me as odd" stream of consciousness :-)

Author: I'm looking into `mdn/data` changes at the moment. It looks like I should update the syntax there.

Author: Adding details in https://github.com/mdn/data/pull/684

> This example uses `contain-intrinsic-size: auto [<length> | none]` with [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility) and offscreen elements.

Reviewer: I'd do it like this - note "demonstrates" because you don't just use those cases.

```suggestion
This example demonstrates `contain-intrinsic-size: auto <length>` and `contain-intrinsic-size: auto none` with [`content-visibility: auto`](/en-US/docs/Web/CSS/content-visibility) and offscreen elements.
```

We can estimate that our paragraphs will be around 500px tall, so the intrinsic size of their parent elements can be set to `500px`.

> When the element is offscreen, the browser will use 500px size for the layout of that element, and when the element needs to be rendered onscreen, the browser will use the computed size of the element and remember this value when the paragraphs are scrolled off and onscreen again.

Reviewer: This describes just the first two divs demonstrated by the example, both of which are only provided for contrast of the things you actually want to demonstrate.

  width: 500px;
}

.auto-length {
  outline: 4px dotted blue;
  content-visibility: auto;
  contain-intrinsic-size: auto 500px;
}

.large-intrinsic-size {
  outline: 4px dotted red;
  /* Setting an inaccurate intrinsic size for the element */
  contain-intrinsic-size: auto 5000px;
}

.small {
  border: 4px dotted green;
  background: lightblue;
  height: 100px;
  width: 100px;
}

.none {
  outline: 4px dotted red;
  contain-intrinsic-size: auto none;
}
```

#### HTML

```html
<div id="container">
  <div class="auto-length">
    <p>Lorem ipsum</p>
  </div>
  <div class="auto-length">
    <p>Lorem ipsum</p>
  </div>
  <div class="auto-length large-intrinsic-size">
    <p class="small">Lorem ipsum</p>
  </div>
  <div class="auto-length large-intrinsic-size">
    <p class="small">Lorem ipsum</p>
  </div>
  <div class="auto-length none">
    <p>Lorem ipsum</p>
  </div>
  <div class="auto-length none">
    <p>Lorem ipsum</p>
  </div>
</div>
```

#### Result

The effect is noticeable in the scrollbar when elements are scrolled into view.
Scrolling is smooth for the first two elements that we have accurate intrinsic sizes for, but the scrollbar jumps when the third element is onscreen.

> The layout shift occurs because the "large-intrinsic-size" elements have content that is around 500px tall, but we have set an inaccurate value of 5000px for its intrinsic size.

Reviewer: I wasn't sure what I was expecting to see from "layout shift" here - i.e. what does that mean. What I think it means is that layout (height in this case) gets calculated assuming the size of hidden elements. Visibly this affects the length of the scroller in the scrollbar (a large hidden view has a smaller scroller widget) and the position of each element in the view as it scrolls.

When something scrolls into view the layout of the element gets recalculated, so the element and its parent sizes change. If this makes the total window size change the scroller widget will change size, and the relative position of the element.

- The first two boxes have an intrinsic size that matches the actual size, so as they flow into view the layout gets recalculated, but does not change.
- The third and fourth box has a huge intrinsic size set, so it makes the initial calculated layout too big. When the box scrolls into view the size is recalculated, making the box and its parent less tall. The effect is that the scroller moves down (we've effectively scrolled further through the box than we'd estimated) and the scroller is longer, because its a smaller scrolled view.
  - These boxes are also much smaller - probably to make the contrast bigger when they scroll into view and to make it obvious that you're now in the box that forces the change. You should say that.
- The last boxes have `auto none`, so they have zero estimated size, so when they scroll into view the size of the parent increases, the scroller decreases in size and moves up the bar.

I think we need to say a bit more of that ^^^ to make what you actually see more obvious alongside what the boxes do.

I tried this first on chrome in which the feature is not supported. Perhaps overkill, but might be cool to color code some of the items using https://developer.mozilla.org/en-US/docs/Web/CSS/@supports if this feature doesn't work on a particular browser and note that too.

Author: >What I think it means is that layout (height in this case) gets calculated assuming the size of hidden elements. Visibly this affects the length of the scroller in the scrollbar (a large hidden view has a smaller scroller widget) and the position of each element in the view as it scrolls.
>
>When something scrolls into view the layout of the element gets recalculated, so the element and its parent sizes change. If this makes the total window size change the scroller widget will change size, and the relative position of the element.

That's it, essentially. The spec also used the term 'placeholder' for the intrinsic size, so that's a good way to think about it. The browser sets aside a placeholder of the size you specify while skipping rendering the element and its children. 

The suggestions are great, incorporating them now.

Author: I've made some improvements in https://github.com/mdn/content/pull/28618/commits/ae80826ec3e6475b8a4ada940b3d32cf7debf6e2 based on your feedback

Author: >Perhaps overkill, but might be cool to color code some of the items using https://developer.mozilla.org/en-US/docs/Web/CSS/@supports if this feature doesn't work on a particular browser and note that too.

Added something in https://github.com/mdn/content/pull/28618/commits/3b8bef04e05f6ad4c63ece662128d2ac1c366f9d but I'm not sure about it
